### PR TITLE
Add musl toolchain building and testing pipeline

### PR DESCRIPTION
**Pull request by illicitonion**
generate-actions.py generates a series of GitHub Actions steps which:
* Build a musl toolchain from source
* Run a bazel build of a sample C++ binary using that toolchain
* Run that C++ binary on the platform it was targeted for
* (Optionally) create a release with those musl toolchains

There are some improvements we can make still (e.g. getting version numbers from tags, allowing adding custom exec_ and target_compatible_with labels to generated toolchains), but I wanted to get the core of this out before iterating on a few of these improvements :)

### CODE REVIEW

File: musl_cc_toolchain_config.bzl [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+    )
+
+    return cc_common.create_cc_toolchain_config_info(
+        ctx = ctx,
+        toolchain_identifier = "k8-musl-toolchain",
+        host_system_name = "local",
+        target_system_name = "local",
+        target_cpu = "k8",
+        target_libc = "musl",
+        compiler = "musl",
```
Comment (fmeum) [inline musl_cc_toolchain_config.bzl:288]: Is this toolchain expected to be compatible with `gcc`? Matching
https://github.com/bazelbuild/rules_cc/blob/c8c38f8c710cbbf834283e4777916b68261b359c/cc/compiler/BUILD#L59 would be good in that case.

File: musl_cc_toolchain_config.bzl [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+            ),
+        ],
+    )
+
+    return cc_common.create_cc_toolchain_config_info(
+        ctx = ctx,
+        toolchain_identifier = "k8-musl-toolchain",
+        host_system_name = "local",
+        target_system_name = "local",
+        target_cpu = "k8",
```
Comment (fmeum) [inline musl_cc_toolchain_config.bzl:286]: I'm not sure whether this is still used, but wouldn't it be incorrect when targeting `arm64`?

File: generate-actions.py [hunk 3]
```diff
@@ -0,0 +1,539 @@
...
+                artifact.source_arch,
+                artifact.source_os,
+                artifact.target_arch,
+            )
+            for artifact in toolchain_infos
+        ]
+    )
+
+    toolchains_to_register = ", ".join(
+        [f'"@musl_toolchains//:{artifact.repo_name}"' for artifact in toolchain_infos]
```
Comment (fmeum) [inline generate-actions.py:290]: Could this be `@musl_toolchains//:all`? If so, this would greatly simplify Bzlmod adoption.

**Comment by illicitonion** [comment]
@fmeum Thanks for the comments, all addressed, PTAL!

**Review by fmeum — APPROVED** [review]

File: .github/workflows/release.yaml [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+    - apple-darwin-x86_64-x86_64
+    steps:
+    - name: Checkout repo
+      uses: actions/checkout@v3
+    - name: Download musl-1.2.3-platform-x86_64-apple-darwin-target-x86_64-linux-musl.tar.gz
+      uses: actions/download-artifact@v3
+      with:
+        name: musl-1.2.3-platform-x86_64-apple-darwin-target-x86_64-linux-musl.tar.gz
+        path: .
+    - name: Download bazelisk as bazel
```
Comment (alexeagle) [inline .github/workflows/release.yaml:97]: nit: every GitHub Actions runner already has bazelisk installed in their base image

File: README.md [hunk 5]
```diff
@@ -0,0 +1,5 @@
+# musl-toolchain
+
+`musl_toolchain` provides a `cc_toolchain` implementation for Bazel.
+
+This toolchain allows cross-compiling binaries for Linux from various platforms. It can be used to produce binaries which don't dynamically link `libc`, by statically linking `musl`'s `libc` implementation.
```
Comment (alexeagle) [inline README.md:5]: Does this mean it's an alternative to aspect-build/gcc-toolchain, grailbio/llvm-toolchain and uber/hermetic-zig-whatever?
Would be nice to answer a yes/no to that here as I think it's the first thing any reader will want to know.

File: deterministic-tar.sh [hunk 6]
```diff
@@ -0,0 +1,37 @@
+#!/bin/bash
+
+set -euo pipefail
```
Comment (alexeagle) [inline deterministic-tar.sh:3]: I guess you can't use the bsdtar toolchain from aspect_bazel_lib?

File: .github/workflows/release.yaml (hunk 4 above)
Comment (illicitonion) [reply .github/workflows/release.yaml:97]: On Linux, we run inside custom docker containers (to isolate ourselves from changing base images), so still need to install.

File: README.md (hunk 5 above)
Comment (illicitonion) [reply README.md:5]: Added some comparisons to the README :)

File: deterministic-tar.sh (hunk 6 above)
Comment (illicitonion) [reply deterministic-tar.sh:3]: Unfortunately this is run outside of Bazel, in the "run `make` in the musl repo" portion - happy to pull in bsdtar if you think that's a reasonably easy thing to do, but using the bazel toolchain itself is probably pretty fiddly.

**Pull request merged**